[FLINK-16194][k8s] Refactor the Kubernetes decorator design

zhengcanbin opened the pull request:
## What is the purpose of the change

So far, Flink has made efforts for the native integration of Kubernetes. However, it is always essential to evaluate the existing design and consider alternatives that have better design and are easier to maintain in the long run. We have suffered from some problems while developing new features base on the current code. Here is some of them:

1. We don’t have a unified monadic-step based orchestrator architecture to construct all the Kubernetes resources.
2. We don’t have dedicated objects or tools for centrally parsing, verifying, and managing the Kubernetes parameters, which has raised some maintenance and inconsistency issues. 

The ultimate goal of this PR is to evolve some of the designs. Here is a summary of the main evolutions.

1. Introduce a unified monadic-step based orchestrator architecture that has a better, cleaner and consistent abstraction for the Kubernetes resources construction process.
2. Introduce some dedicated tools for centrally parsing, verifying, and managing the Kubernetes parameters.

## Open Questions
1. For the new design, we change the owner from the internal Service to the Deployment for GC for some reasons:
- We do not need the internal Service to forward the request from TaskManger to JobManager in the HA mode, that Service would be removed in such a scenario in another issue.
- Things like Deployment are the first citizen in Kubernetes; it is reasonable that one deletes the controller that runs the master leads to clean-up of all the other resources together representing that Application.

2. For the new design, we don't listen for the **ADD** event when creating the rest Service. The previous design assumes that the Service is ready once the client receives the **ADD** event. However, this is incorrect, no matter for the LB or the NodePort type. We plan to open another issue to further discuss and then fix this problem.

## Brief change log

Main changes are:

- [e629bbc](https://github.com/apache/flink/commit/e629bbc4091e9288f74e2d6a9cfd689daabeb4a3) Trivial code clean-up and test code normalization.
- [675151e](https://github.com/apache/flink/commit/675151e02d7b91e0736963ed2b24f2b8c3ff7046): Remove the existing decorator design patterns.
- [0355f0a](https://github.com/apache/flink/commit/0355f0a6d95bca530b4018fecf11db7560626956): Refactor and simplify KubernetesTestBase.
- [edc3d23](https://github.com/apache/flink/commit/edc3d23742d64dcbd07b24b72b674c17ce06b6e7): Remove the Flink Configuration out of KubernetesResource.
- [8d6e520](https://github.com/apache/flink/commit/8d6e5201b336a6238923292f96b9f0563a4f9029): Introduce some dedicated Kubernetes parameters parsing tools.
- [c41a9a2](https://github.com/apache/flink/commit/c41a9a2b5a5f23b820cae038a0611d2d071c4ce9) to [23ed312](https://github.com/apache/flink/commit/23ed31201c0cd08d6bdb715721bba857ead2b520): Introduce the new Kubernetes decorator design pattern.
- [710984c](https://github.com/apache/flink/commit/710984c24f05169a2c1e644676e05dbc573e6c3a): Rework the FlinkKubeClient to employ the new decorator pattern.
- [a47f12d](https://github.com/apache/flink/commit/a47f12d54f832485ad54f273bc5a2f4901d4dce7) to [fb57917](https://github.com/apache/flink/commit/fb57917227853d0477aa1383d399a619146d7170): Minor improvements


## Verifying this change

This PR adds several test classes and many unit tests to catch most of the test branch for the newly decorator design.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

zhengcanbin commented:
@flinkbot run azure

TisonKun commented:
[0001-hotfix-Refactor-subclass-isXXX-to-inherit-methods.txt](https://github.com/apache/flink/files/4263005/0001-hotfix-Refactor-subclass-isXXX-to-inherit-methods.txt)

@zhengcanbin here is a patch for code quality advice that you can take a look at. As we discuss previously it is often better to use override methods instead of control super classes' logic by subclasses' isXXX method.

Generally changes looks good. I will double check the client side code and wait for response from @wangyang0918 and possibly @tillrohrmann .

zhengcanbin commented:
> [0001-hotfix-Refactor-subclass-isXXX-to-inherit-methods.txt](https://github.com/apache/flink/files/4263005/0001-hotfix-Refactor-subclass-isXXX-to-inherit-methods.txt)
> 
> @zhengcanbin here is a patch for code quality advice that you can take a look at. As we discuss previously it is often better to use override methods instead of control super classes' logic by subclasses' isXXX method.
> 
> Generally changes looks good. I will double check the client side code and wait for response from @wangyang0918 and possibly @tillrohrmann .

Hi, tison, thanks for the comment, a new commit [786a505](https://github.com/apache/flink/pull/11233/commits/786a5056c957863c05ad24b00ca1dca032905eb0) resolves the problem.

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -66,8 +70,52 @@ public Fabric8FlinkKubeClient(Configuration flinkConfig, KubernetesClient client
+
+		// create Deployment
+		LOG.debug("Start to create deployment with spec {}", deployment.getSpec().toString());
+		final Deployment createdDeployment = this.internalClient
```
We create the deployment first, then the accompanyingResources(e.g. configmap, service) are created. Does it have chance that the deployment is created with failure? Because the it needs to mount an nonexistent `ConfigMap`.

zhengcanbin replied:
The deployment would never fail since the Kubernetes has strategies for such a case:
1. Generally, the kubelet component helps retry pulling the missing ConfigMap.
2. The controller helps re-deploy the Pod if strategy 1 fails.

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -66,8 +70,52 @@ public Fabric8FlinkKubeClient(Configuration flinkConfig, KubernetesClient client
+			.create(deployment);
+
+		// Note, we should use the server-side uid of the created Deployment for the OwnerReference.
+		setOwnerReference(createdDeployment, accompanyingResources);
```
Since we change the owner reference here, the document needs to be updated.
https://ci.apache.org/projects/flink/flink-docs-master/ops/deployment/native_kubernetes.html#manual-resource-cleanup

zhengcanbin replied:
Fixed by [abfef13](https://github.com/apache/flink/pull/11233/commits/abfef134eb5940d7831dab4c794d654c2fa70263)

wangyang0918 replied:
The e2e test  `test_kubernetes_session.sh` also needs to be updated.

zhengcanbin replied:
Good catch. Fixed by [b088e0e](https://github.com/apache/flink/pull/11233/commits/b088e0e0530fa3178a547b26b7e950165181caaf)

wangyang0918 commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java:
```diff
@@ -47,19 +55,29 @@
+
 	@Test
 	public void testDeploySessionCluster() throws Exception {
+		mockRestServiceWithLoadBalancer(MOCK_SERVICE_HOST_NAME, MOCK_SERVICE_IP);
```
`mockRestServiceWithLoadBalancer` could be removed. It has been called in the `setup`. Other tests is in the same situation.

zhengcanbin replied:
Fixed by [e55c0d2](https://github.com/apache/flink/pull/11233/commits/e55c0d2867d9c79b977bd57997e28ecb2310a616)

wangyang0918 commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java:
```diff
@@ -147,4 +173,30 @@ public void testKillCluster() throws Exception {
+			.always();
+	}
+
+	private Service buildMockRestService(@Nullable String hostname, @Nullable String ip) {
```
`@Nullable` could be removed.

zhengcanbin replied:
Fixed by [d78c1b1](https://github.com/apache/flink/pull/11233/commits/d78c1b1b282a3d7fcfc797bf909dca2b5f2264e4)

wangyang0918 commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java:
```diff
@@ -124,13 +144,19 @@ public void testKillCluster() throws Exception {
 
 		final KubernetesClient kubeClient = server.getClient();
-		assertEquals(2, kubeClient.services().list().getItems().size());
+
```
I am not sure why could not get the rest svc here. Since we just mock the get request, not the post.

zhengcanbin replied:
Actually we mock the request of **ANY** kind, not only the **GET**.

wangyang0918 commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesTestBase.java:
```diff
@@ -18,102 +18,70 @@
+	protected final Configuration flinkConfig = new Configuration();
 
-	protected static final String FLINK_MASTER_ENV_KEY = "LD_LIBRARY_PATH";
+	protected KubernetesClient kubeClient;
```
Since we make `kubeClient` and `flinkKubeClient` as `protected`, these methods `getFabric8FlinkKubeClient(), getKubeClient()` could be removed now.

zhengcanbin replied:
Fixed by [d78c1b1](https://github.com/apache/flink/pull/11233/commits/d78c1b1b282a3d7fcfc797bf909dca2b5f2264e4)

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/KubernetesStepDecorator.java:
```diff
@@ -0,0 +1,47 @@
+
+	/**
+	 * Build the accompanying Kubernetes resources that should be introduced to support this feature. This could
+	 * only applicable on the client-side submission process.
```
That's right.

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/AbstractKubernetesStepDecorator.java:
```diff
@@ -0,0 +1,92 @@
+
+/**
+ * An abstract {@link KubernetesStepDecorator} contains common implementations for different plug-in features
+ * while providing two additional methods.
```
Fixed by [19e32e6](https://github.com/apache/flink/pull/11233/commits/19e32e6bca1a90f608010fdff1236fdfdbb53e3b).

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/ExternalServiceDecorator.java:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * Creates an external Service to expose the ports of the Flink JobManager(s).
+ * This can include the rest port, and the blob server port.
```
Fixed by [0017b8a](https://github.com/apache/flink/pull/11233/commits/0017b8a8930024d0e9ce3355d050565a5007831d)

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/FlinkConfMountDecorator.java:
```diff
@@ -0,0 +1,174 @@
+	String getFlinkConfData(Configuration configuration) throws IOException {
+		try (StringWriter sw = new StringWriter();
+			PrintWriter out = new PrintWriter(sw)) {
+			for (String key : configuration.keySet()) {
```
Fixed by [978c6f0](https://github.com/apache/flink/pull/11233/commits/978c6f00810a66a87325af6454e49e2f084e45be).

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactory.java:
```diff
@@ -0,0 +1,109 @@
+				.endSpec()
+			.build();
+
+		final Map<String, String> labels = resolvedPod.getMetadata().getLabels();
```
The basic logic is:
1. We use the Pod object built by the decorators to fill the pod template of the deployment.
2. Then set the {{deployment.metadata.labels}} separately to help other developers realize that the labels of the deployment could be different from the labels of the pod.

For the usual scenarios, the labels could be the same. For this PR,  resolvedPod.getMetadata().getLabels() and kubernetesJobManagerParameters.getLabels() shares the same values.

zhengcanbin commented:
> @zhengcanbin Thanks a lot for your excellent work. The design and implementation are really good. After this refactor, i believe that adding new features will become easier.
> 
> I have gone with a quick review and leave some comments. I have not tested in a minkube or real k8s cluster. So please hold on and i will give more feedback in this weekend.

Thanks for the review, @wangyang0918.  Kindly expect to see the new feedback by testing this PR in a Minikube or a real k8s cluster.

wangyang0918 commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactoryTest.java:
```diff
@@ -0,0 +1,203 @@
+		assertEquals(1, restServiceCandidates.size());
+
+		final Service resultInternalService = internalServiceCandidates.get(0);
+		assertEquals(2, resultInternalService.getMetadata().getLabels().size());
```
Maybe we could use `getCommonLabels()` to assert the real labels, not only check the size. I think other `getLabels().size()` is in the same situation.

zhengcanbin replied:
Since we have checked the real labels in `InternalServiceDecoratorTest` and `ExternalServiceDecoratorTest`, I don't think it's necessary to check them again in the final test of `KubernetesJobManagerFactoryTest`, rough assert such as size verifying could be enough, otherwise, there are repeated test work.

wangyang0918 commented:
```
echo 'stop' | ./bin/kubernetes-session.sh -Dkubernetes.cluster-id=flink-native-k8s-session-1 -Dexecution.attached=true
```
Probably because we do not include `io.fabric8:zjsonpatch` in the pom.xml of `flink-kubernetes` module.

```
2020-03-02 14:30:32,386 INFO  org.apache.flink.kubernetes.KubernetesClusterDescriptor      [] - Retrieve flink cluster flink-native-k8s-session-1 successfully, JobManager Web Interface: http://11.164.91.5:31318
Exception in thread "main" java.lang.NoClassDefFoundError: io/fabric8/zjsonpatch/JsonDiff
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handlePatch(OperationSupport.java:297)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handlePatch(BaseOperation.java:808)
	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:145)
	at io.fabric8.kubernetes.api.model.apps.DoneableDeployment.done(DoneableDeployment.java:27)
	at io.fabric8.kubernetes.api.model.apps.DoneableDeployment.done(DoneableDeployment.java:6)
	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.patch(HasMetadataOperation.java:151)
	at io.fabric8.kubernetes.client.dsl.internal.RollableScalableResourceOperation.patch(RollableScalableResourceOperation.java:167)
	at io.fabric8.kubernetes.client.dsl.internal.DeploymentOperationsImpl.patch(DeploymentOperationsImpl.java:113)
	at io.fabric8.kubernetes.client.dsl.internal.DeploymentOperationsImpl.patch(DeploymentOperationsImpl.java:45)
	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.lambda$edit$0(HasMetadataOperation.java:53)
	at io.fabric8.kubernetes.api.model.apps.DoneableDeployment.done(DoneableDeployment.java:27)
	at io.fabric8.kubernetes.client.dsl.internal.DeploymentOperationsImpl$DeploymentReaper.reap(DeploymentOperationsImpl.java:245)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.delete(BaseOperation.java:642)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.delete(BaseOperation.java:63)
	at org.apache.flink.kubernetes.kubeclient.Fabric8FlinkKubeClient.stopAndCleanupCluster(Fabric8FlinkKubeClient.java:182)
	at org.apache.flink.kubernetes.KubernetesClusterDescriptor.killCluster(KubernetesClusterDescriptor.java:193)
	at org.apache.flink.kubernetes.KubernetesClusterDescriptor.killCluster(KubernetesClusterDescriptor.java:59)
	at org.apache.flink.kubernetes.cli.KubernetesSessionCli.run(KubernetesSessionCli.java:125)
	at org.apache.flink.kubernetes.cli.KubernetesSessionCli.lambda$main$0(KubernetesSessionCli.java:185)
	at org.apache.flink.runtime.security.contexts.NoOpSecurityContext.runSecured(NoOpSecurityContext.java:30)
	at org.apache.flink.kubernetes.cli.KubernetesSessionCli.main(KubernetesSessionCli.java:185)
Caused by: java.lang.ClassNotFoundException: io.fabric8.zjsonpatch.JsonDiff
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 21 more
```

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/AbstractServiceDecorator.java:
```diff
@@ -0,0 +1,91 @@
+
+	protected abstract String getServiceName();
+
+	protected List<ServicePort> getServicePorts() {
```
I think it's good to keep the current implementation. BTW, these pieces of code would be deleted with another PR for clarifying the design of the so-called internal/external Service.

zhengcanbin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/JavaCmdTaskManagerDecorator.java:
```diff
@@ -0,0 +1,109 @@
+			boolean hasLogback,
+			boolean hasLog4j,
+			String mainClass,
+			@Nullable String mainArgs) {
```
Fixed by [b088e0e](https://github.com/apache/flink/pull/11233/commits/b088e0e0530fa3178a547b26b7e950165181caaf)

zhengcanbin commented:
> ```
> echo 'stop' | ./bin/kubernetes-session.sh -Dkubernetes.cluster-id=flink-native-k8s-session-1 -Dexecution.attached=true
> ```
> 
> Probably because we do not include `io.fabric8:zjsonpatch` in the pom.xml of `flink-kubernetes` module.
> 
> ```
> 2020-03-02 14:30:32,386 INFO  org.apache.flink.kubernetes.KubernetesClusterDescriptor      [] - Retrieve flink cluster flink-native-k8s-session-1 successfully, JobManager Web Interface: http://11.164.91.5:31318
> Exception in thread "main" java.lang.NoClassDefFoundError: io/fabric8/zjsonpatch/JsonDiff
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handlePatch(OperationSupport.java:297)
> 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handlePatch(BaseOperation.java:808)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:145)
> 	at io.fabric8.kubernetes.api.model.apps.DoneableDeployment.done(DoneableDeployment.java:27)
> 	at io.fabric8.kubernetes.api.model.apps.DoneableDeployment.done(DoneableDeployment.java:6)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.patch(HasMetadataOperation.java:151)
> 	at io.fabric8.kubernetes.client.dsl.internal.RollableScalableResourceOperation.patch(RollableScalableResourceOperation.java:167)
> 	at io.fabric8.kubernetes.client.dsl.internal.DeploymentOperationsImpl.patch(DeploymentOperationsImpl.java:113)
> 	at io.fabric8.kubernetes.client.dsl.internal.DeploymentOperationsImpl.patch(DeploymentOperationsImpl.java:45)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.lambda$edit$0(HasMetadataOperation.java:53)
> 	at io.fabric8.kubernetes.api.model.apps.DoneableDeployment.done(DoneableDeployment.java:27)
> 	at io.fabric8.kubernetes.client.dsl.internal.DeploymentOperationsImpl$DeploymentReaper.reap(DeploymentOperationsImpl.java:245)
> 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.delete(BaseOperation.java:642)
> 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.delete(BaseOperation.java:63)
> 	at org.apache.flink.kubernetes.kubeclient.Fabric8FlinkKubeClient.stopAndCleanupCluster(Fabric8FlinkKubeClient.java:182)
> 	at org.apache.flink.kubernetes.KubernetesClusterDescriptor.killCluster(KubernetesClusterDescriptor.java:193)
> 	at org.apache.flink.kubernetes.KubernetesClusterDescriptor.killCluster(KubernetesClusterDescriptor.java:59)
> 	at org.apache.flink.kubernetes.cli.KubernetesSessionCli.run(KubernetesSessionCli.java:125)
> 	at org.apache.flink.kubernetes.cli.KubernetesSessionCli.lambda$main$0(KubernetesSessionCli.java:185)
> 	at org.apache.flink.runtime.security.contexts.NoOpSecurityContext.runSecured(NoOpSecurityContext.java:30)
> 	at org.apache.flink.kubernetes.cli.KubernetesSessionCli.main(KubernetesSessionCli.java:185)
> Caused by: java.lang.ClassNotFoundException: io.fabric8.zjsonpatch.JsonDiff
> 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
> 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
> 	... 21 more
> ```

Good catch. Fixed by [b088e0e](https://github.com/apache/flink/pull/11233/commits/b088e0e0530fa3178a547b26b7e950165181caaf)

zhengcanbin commented:
> @zhengcanbin Thanks a lot for addressing the comments. I have gone though again and the changes look really good to me. I just leave few minor comments.
> 
> For the real K8s cluster tests.
> 
>     * Non-HA session cluster
>       
>       * deploy [SUCCEED]
>       * submit Flink job via `CLI` [SUCCEED]
>       * submit Flink job via webui [SUCCEED]
>       * cancel all jobs and taskmanager released by idle [SUCCEED]
>       * stop via `kubernetes-session.sh` [FAILED]
>       * stop via `kubectl delete deployment/<Cluster_ID>` [SUCCEED]
> 
>     * HA session cluster
>       
>       * Same operations as above [SUCCEED]
>       * Kill taskmanager, job failover and resume successfully [SUCCEED]
>       * Kill jobmanager, job failover and resume successfully [SUCCEED]

@wangyang0918 Thanks for the testing work and all the comments. I have resolved the new comments via commit [b088e0e](https://github.com/apache/flink/pull/11233/commits/b088e0e0530fa3178a547b26b7e950165181caaf).

zhengcanbin commented:
@flinkbot run travis

zhengcanbin commented:
@flinkbot run azure

wangyang0918 commented:
@zhengcanbin Thanks for addressing all the comments. The changes look almost good to me now. Last one comment about the `NOTICE` file. Since we have included a new dependency `zjsonpatch`, it should also be added into `flink-kubernetes/src/main/resources/META-INF/NOTICE`.

zhengcanbin commented:
> @zhengcanbin Thanks for addressing all the comments. The changes look almost good to me now. Last one comment about the `NOTICE` file. Since we have included a new dependency `zjsonpatch`, it should also be added into `flink-kubernetes/src/main/resources/META-INF/NOTICE`.

Fixed by [d77ca1c](https://github.com/apache/flink/commit/d77ca1c8ccec2ec273d3a261c79e8ad1a31633f9).

zhengcanbin commented:
@flinkbot run travis

zhengcanbin commented:
@flinkbot run azure

TisonKun commented:
It seems azure failed on mesos e2e cases. @KarmaGYZ do you have an idea or know an existing issue? (log: https://dev.azure.com/rmetzger/5bd3ef0a-4359-41af-abca-811b04098d2e/_apis/build/builds/5828/logs/118 )

KarmaGYZ commented:
> It seems azure failed on mesos e2e cases. @KarmaGYZ do you have an idea or know an existing issue? (log: https://dev.azure.com/rmetzger/5bd3ef0a-4359-41af-abca-811b04098d2e/_apis/build/builds/5828/logs/118 )

Root cause: FLINK-16288. Please rebase on the latest master branch.

zhengcanbin commented:
> > It seems azure failed on mesos e2e cases. @KarmaGYZ do you have an idea or know an existing issue? (log: https://dev.azure.com/rmetzger/5bd3ef0a-4359-41af-abca-811b04098d2e/_apis/build/builds/5828/logs/118 )
> 
> Root cause: FLINK-16288. Please rebase on the latest master branch.

@KarmaGYZ Thanks for the information.

wangyang0918 commented on flink-kubernetes/src/main/resources/META-INF/NOTICE:
```diff
@@ -15,6 +15,7 @@ This project bundles the following dependencies under the Apache Software Licens
 - com.squareup.okhttp3:okhttp:3.12.1
 - com.squareup.okio:okio:1.15.0
 - io.fabric8:kubernetes-client:4.5.2
+- io.fabric8:zjsonpatch:0.3.0
```
Nit: Maybe we could put the dependencies in alphabetical order.

zhengcanbin replied:
Fixed by [64137c6](https://github.com/apache/flink/pull/11233/commits/64137c6034da191208680bbe89948d911eec6e52)

TisonKun commented:
> @TisonKun Do you think we need to reorganize the commit message or squash them when merging? Currently, i think there are too many `hotfix` to address the comments or previous commits.

Yes it would be nice that if @zhengcanbin can reorganize the commits and squash some intermediate hotfixes. I tend to avoid the PR as a "squash-and-merge" monolithic one or such a series of handy fixups.

Or I will try to reorganize the commits on merging.

TisonKun commented:
```
2020-03-03T04:18:38.2171847Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (integration-tests) on project flink-state-processor-api_2.11: There are test failures.
2020-03-03T04:18:38.2172997Z [ERROR] 
2020-03-03T04:18:38.2173799Z [ERROR] Please refer to /__w/1/s/flink-libraries/flink-state-processing-api/target/surefire-reports for the individual test results.
2020-03-03T04:18:38.2174597Z [ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
2020-03-03T04:18:38.2175072Z [ERROR] ExecutionException Error occurred in starting fork, check output in log
2020-03-03T04:18:38.2175671Z [ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException Error occurred in starting fork, check output in log
2020-03-03T04:18:38.2176787Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter.awaitResultsDone(ForkStarter.java:510)
2020-03-03T04:18:38.2177445Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter.runSuitesForkPerTestSet(ForkStarter.java:457)
2020-03-03T04:18:38.2178053Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:298)
2020-03-03T04:18:38.2178813Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:246)
2020-03-03T04:18:38.2179652Z [ERROR] at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1183)
2020-03-03T04:18:38.2180320Z [ERROR] at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1011)
2020-03-03T04:18:38.2180974Z [ERROR] at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:857)
2020-03-03T04:18:38.2181725Z [ERROR] at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
2020-03-03T04:18:38.2182381Z [ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
2020-03-03T04:18:38.2182907Z [ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
2020-03-03T04:18:38.2183440Z [ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
2020-03-03T04:18:38.2184009Z [ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
2020-03-03T04:18:38.2184645Z [ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
2020-03-03T04:18:38.2185293Z [ERROR] at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
2020-03-03T04:18:38.2185929Z [ERROR] at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
2020-03-03T04:18:38.2186545Z [ERROR] at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
2020-03-03T04:18:38.2187050Z [ERROR] at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
2020-03-03T04:18:38.2187800Z [ERROR] at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
2020-03-03T04:18:38.2188225Z [ERROR] at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
2020-03-03T04:18:38.2188733Z [ERROR] at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
2020-03-03T04:18:38.2189123Z [ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-03-03T04:18:38.2189580Z [ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-03-03T04:18:38.2190104Z [ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-03-03T04:18:38.2190716Z [ERROR] at java.lang.reflect.Method.invoke(Method.java:498)
2020-03-03T04:18:38.2191179Z [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
2020-03-03T04:18:38.2191713Z [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
2020-03-03T04:18:38.2192306Z [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
2020-03-03T04:18:38.2192961Z [ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
2020-03-03T04:18:38.2193642Z [ERROR] Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: Error occurred in starting fork, check output in log
2020-03-03T04:18:38.2194259Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:622)
2020-03-03T04:18:38.2194809Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter.access$600(ForkStarter.java:115)
2020-03-03T04:18:38.2195379Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:444)
2020-03-03T04:18:38.2195921Z [ERROR] at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:420)
2020-03-03T04:18:38.2196512Z [ERROR] at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2020-03-03T04:18:38.2196985Z [ERROR] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2020-03-03T04:18:38.2197519Z [ERROR] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2020-03-03T04:18:38.2197971Z [ERROR] at java.lang.Thread.run(Thread.java:748)
2020-03-03T04:18:38.2198739Z [ERROR] -> [Help 1]
```

azure failed on 'library' profile which doesn't affected by this PR. It seems we failed in irrelevant PRs and master in this case( https://dev.azure.com/rmetzger/Flink/_pipeline/analytics/stageawareoutcome?definitionId=4 ). Is it a known issue? cc @zentol @tzulitai

zhengcanbin commented:
> > @TisonKun Do you think we need to reorganize the commit message or squash them when merging? Currently, i think there are too many `hotfix` to address the comments or previous commits.
> 
> Yes it would be nice that if @zhengcanbin can reorganize the commits and squash some intermediate hotfixes. I tend to avoid merging this PR as a "squash-and-merge" monolithic one or such a series of handy fixups.
> 
> Or I will try to reorganize the commits on merging.

Thanks! I will reorganize the commits later to remove the handy fixups.

TisonKun commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkPod.java:
```diff
@@ -0,0 +1,98 @@
+		return pod;
+	}
+
+	public void setPod(Pod pod) {
```
nit: you can edit the FlinkPod commit to make it a value class, i.e., `pod` and `mainContainer` are `final` and no getters.

TisonKun commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubernetesJobManagerSpecification.java:
```diff
@@ -16,19 +16,32 @@
 
 /**
- * Represent KubernetesDeployment resource in kubernetes.
+ *	Composition of the created Kubernetes components that represents a Flink application.
```
```suggestion
 *  Composition of the created Kubernetes components that represents a Flink application.
```

TisonKun replied:
You can edit the commit introduce this instead of adding fixup. I think we agree on most of codes now.
